DevAcademy/Services/ImageCache.swift

```diff
@@ -0,0 +1,165 @@
+
+        if !contents.contains(hashedURL) {
+            return nil
+        }
+        let data = try? Data(contentsOf: defaultPath.appendingPathComponent(hashedURL))
+        guard let data = data else { return nil }
+        guard let uiimage = UIImage(data: data) else { return nil }
+        return Image(uiImage: uiimage)
+
+        // HINT: Data -> UIImage
+        // let uiimage = UIImage(data: data)
+    }
+
+
+    /// Updates image in the cache.
+    ///
+    /// The function should have following behavior:
+    ///  - Hash the URL
+    ///  - Remove existring file (if exists)
+    ///  - Create a binary data from the image
+    ///  - Save the binary data to the file
+    ///
+    /// - Parameters:
+    ///   - image: Image to be stored
+    ///   - url: The URL that was executed upon the server to get this image.
+    func update(image: UIImage, at url: URL) {
+        let hashedURL: String = hash(of: url)
+        var contents: [String] = try! FileManager.default.contentsOfDirectory(atPath: defaultPath.path(percentEncoded: false))
+
+        if contents.contains(hashedURL) {
+            try! FileManager.default.removeItem(at: defaultPath.appendingPathComponent(hashedURL))
+        }
+        guard let bytes = image.jpegData(compressionQuality: 1.0) else { return }
+        try! bytes.write(to: defaultPath.appendingPathComponent(hashedURL))
+
+
+        // HINT: UIImage -> Data
+        // guard let bytes = image.jpegData(compressionQuality: 1.0) else { return }
+    }
+}
+
+enum StoredAsyncImageError: Error {
+    case decodingFailed
+}
+
+struct StoredAsyncImage<I: View, P: View>: View {
+
+    @State private var image: Image?
+
+    private let url: URL
+    private let imageBuilder: (Image) -> I
+    private let placeholderBuilder: () -> P
+
+
+    init(url: URL, image: @escaping (Image) -> I, placeholder: @escaping () -> P) {
+        self.url = url
+        self.imageBuilder = image
+        self.placeholderBuilder = placeholder
+    }
+
+    private func performURLFetch() async throws -> (UIImage, Image) {
+        let (data, _) = try await URLSession.shared.data(from: url)
+        guard let uiimage = UIImage(data: data) else {
+            throw StoredAsyncImageError.decodingFailed
+        }
+        let image = Image(uiImage: uiimage)
+
+        return (uiimage, image)
+    }
+
+    /// Look into the image cache, whether image is already downloaded.
+    ///
+    /// If so, store it in `image` state variable.
+    /// If not, download the image via `performURLFetch()` function, store it in the cache and in the `image` state vriable.
+    private func loadImage() async {
+        if let image = ImageStorage.shared.loadImage(for: self.url) {
+            self.image = image
+            return
+        }
+        do {
+            let (uiimage, _) = try await performURLFetch()
+            ImageStorage.shared.update(image: uiimage, at: url)
```

Comment:
```suggestion
            let (uiimage, image) = try await performURLFetch()
            ImageStorage.shared.update(image: uiimage, at: url)
            self.image = image
```

Všimněte si rozdílu oproti následujícímu komentáři. Zde definuji `image` na řádku 143. Tím pádem do `self.image` vkládám správný obrázek.